### ISSUE TITLE
Sampling tensor indices by another tensor?

### ISSUE THREAD

**Issue by slerman12** [tensorflow/probability#244]
For some reason, the Distributions tutorial will not load - I have tried in different browsers and I don't know what's wrong. So, I am posting a very naive question here; hopefully it's not too troublesome to answer.

I have a tensor `T` of shape `Batch_Size x Num_Items x Item_Dimension` and another tensor `P` of shape `Batch_Size x Num_Items`, where the Num_Items values in each batch of P sum to 1 (a probability distribution of items for each batch). I want to sample without replacement `N` items from T according to  probability distribution P. The resulting tensor should be of shape `Batch_Size x N x Item_Dimension`. How would I do this?

Thank you for any help.

**Comment by SiegeLordEx** [comment]
The main difficulty here is that we don't support sampling without replacement, as once you generate indices without replacement the actual extraction of values from tensor `T` is a straightforward (if gnarly looking) application of `tf.gather_nd`. Here's a quick and dirty implementation of sampling without replacement and the following gather operation. Note that I'm returning the final tensor with the shape `N x batch x item_dim` because in TFP we prefer to put the sample dimension before the batch dimension.

```python
def sample_without_replacement(categorical, num_samples):

  def body(n, samples, logits):
    dist = tfd.Categorical(logits=logits)
    sample = dist.sample()
    logit_mask = tf.one_hot(sample, tf.shape(logits)[-1], True, False)
    new_logits = tf.where(
        logit_mask, tf.fill(tf.shape(logits), tf.cast(-np.Inf, logits.dtype)),
        logits)
    samples = tf.concat([samples, sample[tf.newaxis, ...]], 0)
    return n + 1, samples, new_logits

  _, samples, logits = tf.while_loop(
      lambda n, *_: n < num_samples,
      body, [
          0,
          tf.zeros(
              tf.concat([[0],
                         categorical.batch_shape_tensor(),
                         categorical.event_shape_tensor()], 0),
              dtype=categorical.dtype), categorical.logits
      ],
      shape_invariants=[
          tf.TensorShape([]),
          tf.TensorShape(None),
          categorical.logits.shape,
      ])
  return samples, tfd.Categorical(logits=logits)


logits = tf.constant([[0.5, 0.2, 0.3], [0.1, 0.2, 0.7]])
data = tf.tile([[[0], [1], [2]], [[3], [4], [5]]], [1, 1, 3])

dist = tfd.Categorical(logits=logits)
samples, _ = sample_without_replacement(dist, 3)

# Once samples are created, you can get the samples from the data like so:
range_indices = tf.tile(
    tf.range(tf.shape(logits)[0])[tf.newaxis, ...], [tf.shape(samples)[0], 1])
data_indices = tf.stack([range_indices, samples], -1)
sampled_data = tf.gather_nd(data, data_indices)

with tf.Session() as sess:
  samples, sampled_data = sess.run([samples, sampled_data])

  print("categorical samples")
  print(samples)
  print(samples.shape)

  print("sampled data")
  print(sampled_data)
  print(sampled_data.shape)
```

Going forward, I think it may make sense to add `sample_without_replacement` to `tfd.Categorical`, with an API similar to what I wrote above (probably with a more efficient implementation though).

**Comment by slerman12** [comment]
Thank you, this helps! There might be a more efficient solution using argmax, but I'm not sure (maybe argmax of a sampled multinomial gaussian with mean at the original logits?). This would probably be an approximation, unless the ideal variance could be determined. I'm just speculating.

**Comment by slerman12** [comment]
I also have a couple extra questions. 

1. Why did you use "logits" when the distribution already sums to 1? 

2. Is there a way to use the result of `dist.sample()` if it is not `None`, otherwise use -1? My items have dynamic lengths, and I would like to use the last (-1th) item, which is just 0-padding, whenever the `dist.sample()` returns `None`. I can't use the current code because the `None`s cause issues with shape inference. A simple fix would be to substitute them with -1. I'm not sure how to check for that since `dist.sample()` returns a `None` of type `_message` I guess.

**Comment by SiegeLordEx** [comment]
That's an interested reference about the Gumbel-Top-K approach, it could probably be faster in your case.

I used logits because after the first sample is generated, the distirbution is no longer normalized and I didn't want to code in re-normalization. Generally, working in log-space for is preferred for probabilities anyway.

I don't fully understand your question about `dist.sample()` returning `None`, that shouldn't ever happen.
